File: src/io/batchify.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 *  Copyright (c) 2020 by Contributors
 * \file batchify.cc
 * \brief Mini-batch data combination functions.
 */
#include <dmlc/parameter.h>
#include <dmlc/omp.h>
#include <mxnet/io.h>
#include <mshadow/tensor.h>
#include <mshadow/extension.h>
#include <mshadow/extension/slice.h>

#include <stack>
#include <cmath>

#include "./inst_vector.h"
#include "../ndarray/ndarray_function.h"

namespace mxnet {
namespace io {


#define tostr(s) #s

#ifdef _MSC_VER
  #if _MSC_VER < 1925
    #define omp_parallel(t) __pragma(omp parallel for num_threads(t))
  #else
    #define omp_parallel(t) _Pragma(tostr(omp parallel for num_threads( ## t ## )))
  #endif
#else
  #define omp_parallel(t) _Pragma(tostr(omp parallel for num_threads(t)))
#endif

struct GroupBatchifyParam : public dmlc::Parameter<GroupBatchifyParam> {
  mxnet::Tuple<std::intptr_t> functions;
  // declare parameters
  DMLC_DECLARE_PARAMETER(GroupBatchifyParam) {
      DMLC_DECLARE_FIELD(functions)
          .describe("Internal sequentially applied batchify functions. "
                    "The number of functions must match output of dataset items.");
  }
};  // struct GroupBatchifyParam
DMLC_REGISTER_PARAMETER(GroupBatchifyParam);

class GroupBatchify : public BatchifyFunction {
 public:
  explicit GroupBatchify(const std::vector<std::pair<std::string, std::string> >& kwargs) {
    param_.InitAllowUnknown(kwargs);
    fs_.reserve(param_.functions.ndim());
    for (int i = 0; i < param_.functions.ndim(); ++i) {
        fs_.emplace_back(*static_cast<BatchifyFunctionPtr*>(
            reinterpret_cast<void*>(param_.functions[i])));
    }
  }

  bool Batchify(const std::vector<std::vector<NDArray> >& inputs,
                        std::vector<NDArray>* outputs) override {
    auto bs = inputs.size();
    CHECK_GT(bs, 0) << "BatchifyFunction should handle at lease 1 sample";
    auto out_size = inputs[0].size();
    CHECK_EQ(out_size, fs_.size()) << "In GroupBatchifyFunction, Elem size "
      << out_size << " and batchify function size " << fs_.size() << " must match";
    outputs->resize(out_size);
    for (size_t i = 0; i < out_size; ++i) {
      std::vector<std::vector<NDArray> > inp;
      inp.reserve(inputs.size());
      for (const auto & input : inputs) {
          std::vector<NDArray> curr({input[i]});
          inp.emplace_back(curr);
      }
      std::vector<NDArray> tmp;
      if (!fs_[i]->Batchify(inp, &tmp)) return false;
      (*outputs)[i] = tmp[0];
    }
    return true;
  }

 private:
  /*! \brief params */
  GroupBatchifyParam param_;
  /*! \brief internal batchify function pointers */
  std::vector<BatchifyFunctionPtr> fs_;
};  // class GroupBatchify

MXNET_REGISTER_IO_BATCHIFY_FUNCTION(GroupBatchify)
  .describe(R"code(Returns the GroupBatchify function.
    )code" ADD_FILELINE)
  .add_arguments(GroupBatchifyParam::__FIELDS__())
  .set_body([](const std::vector<std::pair<std::string, std::string> >& kwargs) {
    return new GroupBatchify(kwargs);
});

struct StackBatchifyParam : public dmlc::Parameter<StackBatchifyParam> {
  /*! \brief Length of the sequence. */
  int use_shared_mem;
  // declare parameters
  DMLC_DECLARE_PARAMETER(StackBatchifyParam) {
      DMLC_DECLARE_FIELD(use_shared_mem).set_default(0)
          .describe("If 1, use shared memory.");
  }
};  // struct StackBatchifyParam

DMLC_REGISTER_PARAMETER(StackBatchifyParam);

class StackBatchify : public BatchifyFunction {
 public:
  explicit StackBatchify(const std::vector<std::pair<std::string, std::string> >& kwargs) {
    param_.InitAllowUnknown(kwargs);
  }

  bool Batchify(const std::vector<std::vector<NDArray> >& inputs,
                        std::vector<NDArray>* outputs) override {
    auto out_size = SanityCheck(inputs);
    auto bs = inputs.size();
    outputs->resize(out_size);
    for (size_t i = 0; i < out_size; ++i) {
        // Process i-th output
        mxnet::TShape ashape = inputs[0][i].shape();
        CHECK_GE(ashape.ndim(), 0) << "Data dim must be larger than 0";
        // check if all shapes are same
        for (size_t j = 1; j < bs; ++j) {
            CHECK_EQ(ashape, inputs[j][i].shape())
              << "StackBatchify requires all data along batch dim to be the same, "
              << "mismatch " << ashape << " vs. " << inputs[j][i].shape();
        }

        // calculate output ndarray size
        TShape sshape(ashape.ndim() + 1, 0);
        sshape[0] = bs;
        for (int k = 0; k < ashape.ndim(); ++k) {
          sshape[k + 1] = ashape[k];
        }

        int dtype = inputs[0][i].dtype();
        if (!(*outputs)[i].is_none() && (*outputs)[i].ctx() == mxnet::Context::CPU(0) &&
            (*outputs)[i].dtype() == dtype &&
            (*outputs)[i].storage_type() == kDefaultStorage) {
          if ((*outputs)[i].shape() != sshape) {
            // realloc
            (*outputs)[i].ReshapeAndAlloc(sshape);
          }
        } else {
          (*outputs)[i] = NDArray(sshape, mxnet::Context::CPU(0), false, inputs[0][i].dtype());
        }
        int sbs = static_cast<int>(bs);
        MSHADOW_TYPE_SWITCH_WITH_BOOL(dtype, DType, {
          omp_parallel(bs)
          for (int j = 0; j < sbs; ++j) {
            omp_exc_.Run([&] {
              // inputs[j][i].WaitToRead();
              DType *ptr = (*outputs)[i].data().dptr<DType>();
              auto asize = ashape.Size();
              RunContext rctx{(*outputs)[i].ctx(), nullptr, nullptr, false};
              auto dst = TBlob(
                ptr + asize * j, inputs[j][i].data().shape_, cpu::kDevMask, dtype, 0);
              mxnet::ndarray::Copy<cpu, cpu>(
                inputs[j][i].data(), &dst, Context::CPU(), Context::CPU(), rctx);
            });
          }
          omp_exc_.Rethrow();
        })
    }
    return true;
  }
 private:
  /*! \brief parameters */
  StackBatchifyParam param_;
  /*! \brief OMPException obj to store and rethrow exceptions from omp blocks*/
  dmlc::OMPException omp_exc_;

  std::size_t SanityCheck(const std::vector<std::vector<NDArray> >& inputs) {
    auto bs = inputs.size();
    CHECK_GT(bs, 0) << "BatchifyFunction should handle at lease 1 sample";
    auto out_size = inputs[0].size();
    // sanity check: each input has same size
    for (size_t i = 1; i < bs; ++i) {
        CHECK_EQ(inputs[i].size(), out_size)
          << i << "-th input size does not match " << out_size;
    }
    return out_size;
  }
};  // class StackBatchify

MXNET_REGISTER_IO_BATCHIFY_FUNCTION(StackBatchify)
  .describe(R"code(Returns the StackBatchify function.
    )code" ADD_FILELINE)
  .add_arguments(StackBatchifyParam::__FIELDS__())
  .set_body([](const std::vector<std::pair<std::string, std::string> >& kwargs) {
    return new StackBatchify(kwargs);
});

struct PadBatchifyParam : public dmlc::Parameter<PadBatchifyParam> {
  int use_shared_mem;
  double pad_val;
  int dtype;
  int round_to;
  // declare parameters
  DMLC_DECLARE_PARAMETER(PadBatchifyParam) {
      DMLC_DECLARE_FIELD(use_shared_mem).set_default(0)
          .describe("If 1, use shared memory.");
      DMLC_DECLARE_FIELD(pad_val).set_default(0)
          .describe("The filled values, default to 0.");
      DMLC_DECLARE_FIELD(dtype).set_default(-1)
        .describe("If not -1, force to use dtype as output type, otherwise use input type.");
      DMLC_DECLARE_FIELD(round_to).set_default(-1)
        .describe("If > 0, the padded dimension will be rounded to be multiple of this value.");
  }
};  // struct PadBatchifyParam

DMLC_REGISTER_PARAMETER(PadBatchifyParam);

class PadBatchify : public BatchifyFunction {
 public:
  explicit PadBatchify(const std::vector<std::pair<std::string, std::string> >& kwargs) {
    param_.InitAllowUnknown(kwargs);
  }

  bool Batchify(const std::vector<std::vector<NDArray> >& inputs,
                        std::vector<NDArray>* outputs) override {
    auto bs = inputs.size();
    CHECK_GT(bs, 0) << "BatchifyFunction should handle at lease 1 sample";
    auto out_size = inputs[0].size();
    outputs->resize(out_size);
    for (size_t i = 0; i < out_size; ++i) {
        // Process i-th output
        mxnet::TShape ashape = inputs[0][i].shape();
        CHECK_GE(ashape.ndim(), 0) << "Data dim must be larger than 0";
        // find the maximum size in each dim
        for (size_t j = 1; j < bs; ++j) {
          mxnet::TShape other_shape = inputs[j][i].shape();
          CHECK_EQ(ashape.ndim(), other_shape.ndim())
            << "PadBatchify expects all inputs to have same dimensionality: given "
            << ashape.ndim() << " vs. " << other_shape.ndim();
            for (dim_t k = 0; k < ashape.ndim(); ++k) {
              ashape[k] = std::max(ashape[k], other_shape[k]);
            }
        }
        for (dim_t k = 0; k < ashape.ndim(); ++k) {
          // pad to multiple of round_to
          if (param_.round_to > 0) {
            ashape[k] = param_.round_to * static_cast<int>(
              std::ceil(static_cast<double>(ashape[k] / param_.round_to)));
          }
        }

        // calculate output ndarray size
        TShape sshape(ashape.ndim() + 1, 0);
        sshape[0] = bs;
        for (int k = 0; k < ashape.ndim(); ++k) {
          sshape[k + 1] = ashape[k];
        }

        int dtype = param_.dtype > -1 ? param_.dtype : inputs[0][i].dtype();
        if (!(*outputs)[i].is_none() &&
            (*outputs)[i].ctx() == mxnet::Context::CPU(0) &&
            (*outputs)[i].dtype() == dtype &&
            (*outputs)[i].storage_type() == kDefaultStorage) {
          if ((*outputs)[i].shape() != sshape) {
            // realloc
            (*outputs)[i].ReshapeAndAlloc(sshape);
          }
        } else {
          (*outputs)[i] = NDArray(sshape, mxnet::Context::CPU(0), false, inputs[0][i].dtype());
        }
        MSHADOW_TYPE_SWITCH_WITH_BOOL(dtype, DType, {
          // fill pad value first
          std::fill((*outputs)[i].data().dptr<DType>(),
                    (*outputs)[i].data().dptr<DType>() + sshape.Size(),
                    static_cast<DType>(param_.pad_val));
          DType *ptr = (*outputs)[i].data().dptr<DType>();
          auto asize = ashape.Size();
          int sbs = static_cast<int>(bs);
          omp_parallel(bs)
          for (int j = 0; j < sbs; ++j) {
            using namespace mshadow::expr;
            auto compact_shapes = CompactShapes(ashape, inputs[j][i].shape());
            // inputs[j][i].WaitToRead();
            auto& fshape = compact_shapes.first;
            auto& cshape = compact_shapes.second;
            switch (fshape.size()) {
              case 1U: {
                mshadow::Tensor<cpu, 1, DType> dst = TBlob(
                  ptr + asize * j, ashape, cpu::kDevMask, dtype, 0).get_with_shape<cpu, 1, DType>(
                    mshadow::Shape1(fshape[0]));
                mshadow::Tensor<cpu, 1, DType> src = inputs[j][i].data().get_with_shape<
                  cpu, 1, DType>(mshadow::Shape1(cshape[0]));
                slice<0>(dst, 0, cshape[0]) = src;
                break;
              }
              case 2U: {
                mshadow::Tensor<cpu, 2, DType> dst = TBlob(
                  ptr + asize * j, ashape, cpu::kDevMask, dtype, 0).get_with_shape<cpu, 2, DType>(
                    mshadow::Shape2(fshape[0], fshape[1]));
                mshadow::Tensor<cpu, 2, DType> src = inputs[j][i].data().get_with_shape<
                  cpu, 2, DType>(mshadow::Shape2(cshape[0], cshape[1]));
                slice<1>(slice<0>(dst, 0, cshape[0]), 0, cshape[1]) = src;
                break;
              }
              case 3U: {
                mshadow::Tensor<cpu, 3, DType> dst = TBlob(
                  ptr + asize * j, ashape, cpu::kDevMask, dtype, 0).get_with_shape<cpu, 3, DType>(
                    mshadow::Shape3(fshape[0], fshape[1], fshape[2]));
                mshadow::Tensor<cpu, 3, DType> src = inputs[j][i].data().get_with_shape<
                  cpu, 3, DType>(mshadow::Shape3(cshape[0], cshape[1], cshape[2]));
                slice<2>(slice<1>(slice<0>(dst, 0, cshape[0]), 0, cshape[1]), 0, cshape[2]) = src;
                break;
              }
              case 4U: {
                mshadow::Tensor<cpu, 4, DType> dst = TBlob(
                  ptr + asize * j, ashape, cpu::kDevMask, dtype, 0).get_with_shape<cpu, 4, DType>(
                    mshadow::Shape4(fshape[0], fshape[1], fshape[2], fshape[3]));
                mshadow::Tensor<cpu, 4, DType> src = inputs[j][i].data().get_with_shape<
                  cpu, 4, DType>(mshadow::Shape4(cshape[0], cshape[1], cshape[2], cshape[3]));
                slice<3>(slice<2>(slice<1>(slice<0>(dst, 0, cshape[0]), 0, cshape[1]),
                                                    0, cshape[2]), 0, cshape[3]) = src;
                break;
              }
              case 5U: {
                mshadow::Tensor<cpu, 5, DType> dst = TBlob(
                  ptr + asize * j, ashape, cpu::kDevMask, dtype, 0).get_with_shape<cpu, 5, DType>(
                    mshadow::Shape5(fshape[0], fshape[1], fshape[2], fshape[3], fshape[4]));
                mshadow::Tensor<cpu, 5, DType> src = inputs[j][i].data().get_with_shape<
                  cpu, 5, DType>(mshadow::Shape5(
                    cshape[0], cshape[1], cshape[2], cshape[3], cshape[4]));
                slice<4>(slice<3>(slice<2>(slice<1>(slice<0>(
                  dst, 0, cshape[0]), 0, cshape[1]), 0, cshape[2]),
                  0, cshape[3]), 0, cshape[4]) = src;
                break;
              }
              default: {
                LOG(FATAL) << "# dim to pad: " << cshape.size() << " exceeds limit of 5.";
              }
            }
          }
        })
    }
    return true;
  }

 private:
  /*! \brief parameters */
  PadBatchifyParam param_;
  /*! \brief OMPException obj to store and rethrow exceptions from omp blocks*/
  dmlc::OMPException omp_exc_;

  std::pair<std::vector<dim_t>, std::vector<dim_t>> CompactShapes(const TShape& ashape,
                                                                  const TShape& ishape) {
    // squeeze dimensions that do not need pad
    std::stack<dim_t> dim_stack;
    std::vector<dim_t> full_shape;
    std::vector<dim_t> data_shape;
    for (dim_t k = 0; k < ishape.ndim(); ++k) {
      if (ishape[k] == ashape[k]) {
        dim_stack.push(ishape[k]);
      } else {
        dim_t ss = 1;
        while (!dim_stack.empty()) {
          ss *= dim_stack.top();
          dim_stack.pop();
        }
        if (ss > 1) {
          full_shape.emplace_back(ss);
          data_shape.emplace_back(ss);
        }
        full_shape.emplace_back(ashape[k]);
        data_shape.emplace_back(ishape[k]);
      }
    }
    // clear the stack
    index_t ss = 1;
    while (!dim_stack.empty()) {
      ss *= dim_stack.top();
      dim_stack.pop();
    }
    if (ss > 1 || full_shape.empty()) {
      full_shape.emplace_back(ss);
      data_shape.emplace_back(ss);
    }
    CHECK_EQ(full_shape.size(), data_shape.size());
    CHECK_GE(data_shape.size(), 1U);
    return std::make_pair(full_shape, data_shape);
  }
};  // class PadBatchify

MXNET_REGISTER_IO_BATCHIFY_FUNCTION(PadBatchify)
  .describe(R"code(Returns the StackBatchify function.
    )code" ADD_FILELINE)
  .add_arguments(PadBatchifyParam::__FIELDS__())
  .set_body([](const std::vector<std::pair<std::string, std::string> >& kwargs) {
    return new PadBatchify(kwargs);
});
}  // namespace io
}  // namespace mxnet

File: python/mxnet/gluon/data/batchify.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

# coding: utf-8
# pylint: disable=reimported, consider-using-enumerate
"""Batchify function."""
import math
import warnings
import numpy as np

from ...device import Device, cpu
from ... import ndarray as nd
from ... import numpy as _np
from ...util import is_np_array

class Stack(object):
    r"""Stack the input data samples to construct the batch.
    The N input samples must have the same shape/length and will be stacked to construct a batch.
    Examples
    --------
    >>> from mxnet.gluon.data import batchify
    >>> # Stack multiple lists
    >>> a = [1, 2, 3, 4]
    >>> b = [4, 5, 6, 8]
    >>> c = [8, 9, 1, 2]
    >>> batchify.Stack()([a, b, c])
    [[1. 2. 3. 4.]
     [4. 5. 6. 8.]
     [8. 9. 1. 2.]]
    <NDArray 3x4 @cpu(0)>
    >>> # Stack multiple numpy.ndarrays
    >>> import numpy as np
    >>> a = np.array([[1, 2, 3, 4], [5, 6, 7, 8]])
    >>> b = np.array([[5, 6, 7, 8], [1, 2, 3, 4]])
    >>> batchify.Stack()([a, b])
    [[[1. 2. 3. 4.]
      [5. 6. 7. 8.]]
     [[5. 6. 7. 8.]
      [1. 2. 3. 4.]]]
    <NDArray 2x2x4 @cpu(0)>
    >>> # Stack multiple NDArrays
    >>> import mxnet as mx
    >>> a = nd.array([[1, 2, 3, 4], [5, 6, 7, 8]])
    >>> b = nd.array([[5, 6, 7, 8], [1, 2, 3, 4]])
    >>> batchify.Stack()([a, b])
    [[[1. 2. 3. 4.]
      [5. 6. 7. 8.]]
     [[5. 6. 7. 8.]
      [1. 2. 3. 4.]]]
    <NDArray 2x2x4 @cpu(0)>
    """
    def __init__(self, use_shared_mem=False):
        self._use_shared_mem = use_shared_mem

    def __call__(self, data):
        """Batchify the input data
        Parameters
        ----------
        data : list
            The input data samples
        Returns
        -------
        batch_data : NDArray
        """
        _arr = _np if is_np_array() else nd
        _arr_cls = _arr.ndarray if is_np_array() else _arr.NDArray
        if isinstance(data[0], _arr_cls):
            dtype = data[0].dtype
            if self._use_shared_mem:
                out = _arr.empty((len(data),) + data[0].shape, dtype=dtype,
                                 ctx=Device('cpu_shared', 0))
                return _arr.stack(data, out=out) if is_np_array() else _arr.stack(*data, out=out)
            else:
                return _arr.stack(data) if is_np_array() else _arr.stack(*data)
        elif isinstance(data[0], (tuple, list)):
            data = zip(*data)
            return [self.__call__(i) for i in data]
        else:
            out = np.asarray(data)
            dtype = out.dtype
            if self._use_shared_mem:
                return _arr.array(out, ctx=Device('cpu_shared', 0), dtype=dtype)
            else:
                return _arr.array(out, dtype=dtype)

    def __mx_handle__(self):
        from ._internal import StackBatchify
        return StackBatchify()

def _pad_arrs_to_max_length(arrs, pad_val, use_shared_mem, dtype, round_to=None):
    """Inner Implementation of the Pad batchify
    Parameters
    ----------
    arrs : list
    pad_val : number
    use_shared_mem : bool, default False
    round_to : int

    Returns
    -------
    ret : NDArray
    """
    _arr = _np if is_np_array() else nd
    _arr_cls = _np.ndarray if is_np_array() else nd.NDArray
    if isinstance(arrs[0], _arr_cls):
        dtype = arrs[0].dtype if dtype is None else dtype
        arrs = [arr.asnumpy() for arr in arrs]
    elif not isinstance(arrs[0], np.ndarray):
        arrs = [np.asarray(ele) for ele in arrs]
        dtype = arrs[0][0].dtype if dtype is None else dtype
    else:
        dtype = arrs[0].dtype if dtype is None else dtype

    ret_shape = list(arrs[0].shape)
    for pad_axis in range(len(ret_shape)):
        curr_lengths = [ele.shape[pad_axis] for ele in arrs]
        max_size = max(curr_lengths)
        if round_to is not None:
            max_size = round_to * math.ceil(max_size / round_to)
        ret_shape[pad_axis] = max_size
    ret_shape = (len(arrs), ) + tuple(ret_shape)

    ret = np.full(shape=ret_shape, fill_value=pad_val, dtype=dtype)

    for i, arr in enumerate(arrs):
        if arr.shape == ret_shape[1:]:
            ret[i] = arr
        else:
            slices = [slice(None) for _ in range(arr.ndim)]
            for pad_axis in range(arr.ndim):
                slices[pad_axis] = slice(0, arr.shape[pad_axis])
                assert slices[pad_axis].start != slices[pad_axis].stop
            slices = [slice(i, i + 1)] + slices
            ret[tuple(slices)] = arr


    device = Device('cpu_shared', 0) if use_shared_mem else cpu()
    ret = _arr.array(ret, ctx=device, dtype=dtype)

    return ret


class Pad(object):
    """Pad the input ndarrays along the specific padding axis and stack them to get the output.
    Input of the function will be N samples. Each sample should contain a single element that
    can be 1) numpy.ndarray, 2) mxnet.nd.NDArray, 3) list of numbers.
    You can set the `pad_val` to determine the padding value.

    The arrays will be padded to the largest dimensions(at most 5 dimensions to pad) and then
    stacked to form the final output.

    Parameters
    ----------
    val : float or int, default None
        The padding value.
    dtype : str or numpy.dtype, default None
        The value type of the output. If it is set to None, the input data type is used.
    round_to : int, default None
        If specified, the padded dimension will be rounded to be multiple of this argument.

    Examples
    --------
    >>> from mxnet.gluon.data import batchify
    >>> # Inputs are multiple lists
    >>> a = [1, 2, 3, 4]
    >>> b = [4, 5, 6]
    >>> c = [8, 2]
    >>> batchify.Pad()([a, b, c])
    [[ 1  2  3  4]
     [ 4  5  6  0]
     [ 8  2  0  0]]
    <NDArray 3x4 @cpu(0)>
    >>> # Also output the lengths
    >>> a = [1, 2, 3, 4]
    >>> b = [4, 5, 6]
    >>> c = [8, 2]
    >>> # Inputs are multiple ndarrays
    >>> import numpy as np
    >>> a = np.array([[1, 2, 3, 4], [5, 6, 7, 8]])
    >>> b = np.array([[5, 8], [1, 2]])
    >>> batchify.Pad(val=-1)([a, b])
    [[[ 1  2  3  4]
      [ 5  6  7  8]]
     [[ 5  8 -1 -1]
      [ 1  2 -1 -1]]]
    <NDArray 2x2x4 @cpu(0)>
    >>> # Inputs are multiple NDArrays
    >>> import mxnet as mx
    >>> a = nd.array([[1, 2, 3, 4], [5, 6, 7, 8]])
    >>> b = nd.array([[5, 8], [1, 2]])
    >>> batchify.Pad(val=-1)([a, b])
    [[[ 1.  2.  3.  4.]
      [ 5.  6.  7.  8.]]
     [[ 5.  8. -1. -1.]
      [ 1.  2. -1. -1.]]]
    <NDArray 2x2x4 @cpu(0)>
    """
    def __init__(self, val=None, dtype=None, round_to=None, use_shared_mem=False):
        self._pad_val = 0 if val is None else val
        self._dtype = dtype
        self._warned = False
        self._round_to = round_to
        self._use_shared_mem = use_shared_mem

    def __call__(self, data):
        """Batchify the input data.

        The input can be list of numpy.ndarray, list of numbers or list of
        mxnet.nd.NDArray. Inputting mxnet.nd.NDArray is discouraged as each
        array need to be converted to numpy for efficient padding.
        The arrays will be padded to the largest dimension at `axis` and then
        stacked to form the final output.

        Parameters
        ----------
        data : List[np.ndarray] or List[List[dtype]] or List[nd.NDArray]
            List of samples to pad and stack.
        Returns
        -------
        batch_data: NDArray
            Data in the minibatch. Shape is (N, ...)
        """
        _arr = _np if is_np_array() else nd
        _arr_cls = _arr.ndarray if is_np_array() else _arr.NDArray
        if isinstance(data[0], _arr_cls) and not self._warned:
            self._warned = True
            warnings.warn(
                'Using Pad with NDArrays is discouraged for speed reasons. '
                'Instead you should pad your data while it is still a list '
                'and before converting to an NDArray. '
                'Alternatively you can consider inputting a numpy.ndarray.')
        if isinstance(data[0], (_arr_cls, np.ndarray, list)):
            padded_arr = _pad_arrs_to_max_length(data, self._pad_val,
                                                 self._use_shared_mem,
                                                 self._dtype, self._round_to)
            return padded_arr
        else:
            raise NotImplementedError(
                "Pad() does not support multiple items, use Group(Pad(), Pad(), ...) instead")

    def __mx_handle__(self):
        from ._internal import PadBatchify
        return PadBatchify(pad_val=self._pad_val, dtype=self._dtype if self._dtype is not None else -1)

def _append_arrs(arrs, use_shared_mem=False, expand=False, batch_axis=0):
    """Internal impl for returning appened arrays as list."""
    _arr = _np if is_np_array() else nd
    if isinstance(arrs[0], _arr.NDArray):
        if use_shared_mem:
            out = [x.as_in_context(Device('cpu_shared', 0)) for x in arrs]
        else:
            out = arrs
    else:
        if use_shared_mem:
            out = [_arr.array(x, ctx=Device('cpu_shared', 0)) for x in arrs]
        else:
            out = [_arr.array(x) for x in arrs]

    # add batch axis
    if expand:
        out = [x.expand_dims(axis=batch_axis) for x in out]
    return out


class Append(object):
    r"""Loosely return list of the input data samples.
    There is no constraint of shape for any of the input samples, however, you will
    only be able to apply single batch operations since the output have different shapes.
    Examples
    --------
    >>> a = [1, 2, 3, 4]
    >>> b = [4, 5, 6]
    >>> c = [8, 2]
    >>> batchify.Append()([a, b, c])
    [
    [[1. 2. 3. 4.]]
    <NDArray 1x4 @cpu_shared(0)>,
    [[4. 5. 6.]]
    <NDArray 1x3 @cpu_shared(0)>,
    [[8. 2.]]
    <NDArray 1x2 @cpu_shared(0)>
    ]
    """

    def __init__(self, expand=True, batch_axis=0, use_shared_mem=False):
        self._expand = expand
        self._batch_axis = batch_axis
        self._use_shared_mem = use_shared_mem

    def __call__(self, data):
        """Batchify the input data.
        Parameters
        ----------
        data : list
            The input data samples
        Returns
        -------
        batch_data : NDArray
        """
        return _append_arrs(data, use_shared_mem=self._use_shared_mem,
                            expand=self._expand, batch_axis=self._batch_axis)

class Group(object):
    """Wrap multiple batchify functions together. The input functions will be applied
    to the corresponding input fields.
    Each data sample should be a list or tuple containing multiple attributes. The `i`th batchify
    function stored in `Group` will be applied on the `i`th attribute. For example, each
    data sample is (nd_data, label). You can wrap two batchify functions using
    `Group(DataBatchify, LabelBatchify)` to batchify nd_data and label correspondingly.
    Parameters
    ----------
    fn : list or tuple or callable
        The batchify functions to wrap.
    *args : tuple of callable
        The additional batchify functions to wrap.
    Examples
    --------
    >>> a = ([1, 2, 3, 4], 0)
    >>> b = ([5, 7], 1)
    >>> c = ([1, 2, 3, 4, 5, 6, 7], 0)
    >>> f1, f2 = Group(Pad(val=0),
    ...                Stack())([a, b])
    >>> f1
    <BLANKLINE>
    [[1. 2. 3. 4.]
     [5. 7. 0. 0.]]
    <NDArray 2x4 @cpu_shared(0)>
    >>> f2
    <BLANKLINE>
    [0 1]
    <NDArray 2 @cpu_shared(0)>
    """
    def __init__(self, fn, *args):
        self._handle = None
        if isinstance(fn, (list, tuple)):
            assert len(args) == 0, 'Input pattern not understood. The input of Group can be ' \
                                   'Group(A, B, C) or Group([A, B, C]) or Group((A, B, C)). ' \
                                   f'Received fn={str(fn)}, args={str(args)}'
            self._fn = fn
        else:
            self._fn = (fn, ) + args
        for i, ele_fn in enumerate(self._fn):
            assert hasattr(ele_fn, '__call__'), 'Batchify functions must be callable! ' \
                                                f'type(fn[{i}]) = {str(type(ele_fn))}'

    def __call__(self, data):
        """Batchify the input data.
        Parameters
        ----------
        data : list
            The samples to batchfy. Each sample should contain N attributes.
        Returns
        -------
        ret : tuple
            A tuple of length N. Contains the batchified result of each attribute in the input.
        """
        assert len(data[0]) == len(self._fn),\
            'The number of attributes in each data sample should contains' \
            ' {} elements'.format(len(self._fn))
        ret = []
        for i, ele_fn in enumerate(self._fn):
            ret.append(ele_fn([ele[i] for ele in data]))
        return tuple(ret)

    def __mx_handle__(self):
        if self._handle  is None:
            from ._internal import GroupBatchify
            try:
                mx_fn = [fn.__mx_handle__() for fn in self._fn]
                self._handle = GroupBatchify(functions=mx_fn)
            except Exception as e:
                raise NotImplementedError(
                    "GroupBatchify requires all internal batchify functions supported by backend."
                    + str(e))
        return self._handle

class AsList(object):
    """Simply forward the list of input data.
    This is particularly useful when the Dataset contains textual data
    and in conjonction with the `Group` batchify function.
    Examples
    --------
    >>> a = ([1, 2, 3, 4], "I am using MXNet")
    >>> b = ([5, 7, 2, 5], "Gluon rocks!")
    >>> c = ([1, 2, 3, 4], "Batchification!")
    >>> _, l = Group(Stack(), AsList())([a, b, c])
    >>> l
    ['I am using MXNet', 'Gluon rocks!', 'Batchification!']
    """
    def __call__(self, data):
        """
        Parameters
        ----------
        data : list
            The list of samples
        Returns
        -------
        ret : list
            The input list
        """
        return list(data)
